// @sap-ux/fe-mockserver-core/dist/plugins/fileSystemLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graceful_fs_1 = require("graceful-fs");
const util_1 = require("util");
const readFileP = (0, util_1.promisify)(graceful_fs_1.readFile);
const accessP = (0, util_1.promisify)(graceful_fs_1.access);
class FileSystemLoader {
    async loadFile(filePath) {
        return readFileP(filePath, 'utf-8');
    }
    async exists(filePath) {
        try {
            await accessP(filePath);
            return true;
        }
        catch {
            return false;
        }
    }
    syncSupported() {
        return true;
    }
    existsSync(filePath) {
        try {
            (0, graceful_fs_1.accessSync)(filePath);
            return true;
        }
        catch {
            return false;
        }
    }
    loadFileSync(filePath) {
        return (0, graceful_fs_1.readFileSync)(filePath, 'utf-8');
    }
    async loadJS(filePath) {
        delete require.cache[filePath];
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        return Promise.resolve(require(filePath));
    }
}
exports.default = FileSystemLoader;
